File: exercises/Fortran90/PingPong/pingpong.f90
!
! Program in which 2 processes repeatedly pass a message back and forth
!
! The same data is sent from A to B, then returned from B to A.
!

program pingpong
implicit none
include 'mpif.h'

integer :: ierr, size, rank, comm, i, length, numiter
integer :: status(MPI_STATUS_SIZE)
integer :: tag1, tag2, extent
character*10 temp_char10
integer :: iargc
real, allocatable :: sbuffer(:)
double precision :: tstart, tstop, time, totmess

comm = MPI_COMM_WORLD
tag1 = 1
tag2 = 2

call MPI_INIT(ierr)
call MPI_COMM_RANK(comm,rank,ierr)
call MPI_COMM_SIZE(comm,size,ierr)

if (iargc() /= 2) then
 if (rank .eq. 0) then
   write(*,*) 'Usage: pingpong <array length> <number of iterations>'
 end if

 call mpi_finalize(ierr)
 stop
end if


if (rank.gt.1) then
 print*, 'Rank not participating', rank
end if


if (rank .eq. 0) then 
 call getarg(1,temp_char10)
 read(temp_char10,*) length
 call getarg(2,temp_char10)
 read(temp_char10,*) numiter

 print*, 'Array length, number of iterations = '
 print*,  length, numiter
end if

call MPI_BCAST(length,1,MPI_INTEGER,0,comm,ierr)
call MPI_BCAST(numiter,1,MPI_INTEGER,0,comm,ierr)

! Must be run on at least 2 processors
if(size.lt.2)then
 if(rank.eq.0) write(*,*) ' The code must be run on at least 2 processors.'
 call MPI_FINALIZE(ierr)
 stop
endif

! Allocate array
allocate(sbuffer(length))

! Send 'buffer' back and forth between rank 0 and rank 1.
do i=1,length
 sbuffer(i) = rank + 10.d0
enddo

! Start timing the parallel part here.
call MPI_BARRIER(comm,ierr)                                           
tstart = MPI_Wtime()

do i=1,numiter
 if (rank.eq.0)then
  call MPI_SSEND(sbuffer(1),length,MPI_REAL,1,tag1,comm,ierr)
  call MPI_RECV(sbuffer(1),length,MPI_REAL,1,tag2,comm,status,ierr)

 else if (rank.eq.1)then
  call MPI_RECV(sbuffer(1),length,MPI_REAL,0,tag1,comm,status,ierr)
  call MPI_SSEND(sbuffer(1),length,MPI_REAL,0,tag2,comm,ierr)
 endif
enddo


tstop = MPI_Wtime()
time  = tstop - tstart

call MPI_TYPE_SIZE(MPI_REAL,extent,ierr)

if(rank.eq.0)then
 totmess = 2.d0*extent*length/1024.d0*numiter/1024.d0
 write(*,*) ' Ping-Pong of twice ',extent*length,' bytes, for ',numiter,' times.'
 write(*,*) 'Total computing time is ',time,' [s].'
 write(*,*) 'Total message size is ',totmess,' [MB].'
 write(*,*) 'Latency (time per message) is ', time/numiter*0.5d0,'[s].'
 write(*,*) 'Bandwidth (message per time) is ',totmess/time,' [MB/s].'

 if(time.lt.1.d0)then
  write(*,*) 'WARNING: The time is too short to be meaningful, increase the number of iterations and/or the array size so time is at least one second!'
    
    
 endif
endif

deallocate(sbuffer)

call MPI_FINALIZE(ierr)

end program pingpong

File: exercises/C/PingPong/pingpong.c
//
//  Program in which 2 processes repeatedly pass a message back and forth
//
// The same data is sent from A to B, then returned from B to A.
//
// Corrected July 2010 ($HOME/mail/EPCC/NAIS/DSH)
//

#include <stdlib.h>
#include <stdio.h>
#include <mpi.h>

int main(int argc, char* argv[]){
  int rank, size, numiter,length;
  int start,stop,i,iter;
  int tag1, tag2,extent;
  double *sbuffer;
  double t1, t2,totmess;
  MPI_Comm comm;
  MPI_Status status;

  comm  = MPI_COMM_WORLD;
  tag1= 1;
  tag2 = 2;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(comm, &rank);            
  MPI_Comm_size(comm, &size); 

  // Abort if run on less than 2 processors.
  if(size < 2){
    if(rank == 0){
      printf("The code must be run on at least 2 processors.\n");
    }
    MPI_Finalize();
    exit(1);
  }
  
  if(argc < 2) {
    if(rank == 0){
     printf("Code requires 2 input arguments: The array length and the number of iterations. \n");
    }
    MPI_Finalize();
  exit(1);
  }

  if(rank > 1){
    printf( "Rank %d not participating \n",rank);
  }

  if (rank == 0) {
    length = atoi(argv[1]);
    numiter = atoi(argv[2]);
    printf("Array length = %d , number of iterations = %d\n",length,numiter);
  }

  MPI_Bcast(&length,1,MPI_INT,0,comm);
  MPI_Bcast(&numiter,1,MPI_INT,0,comm);


  // Allocate array
  sbuffer = malloc(length*sizeof(double));
 
  // Send 'buffer' back and forth between rank 0 and rank 1.
  for(i=0;i<length;i++)
    sbuffer[i] = (double)rank + 10.0;

  // Time the parallel execution.
  MPI_Barrier(comm);
  t1 = MPI_Wtime();

  // Swap back and forth for iter times

  for(iter=1;iter<numiter;iter++){

  if(rank == 0){

  // Timings only make sense when no print statements are includes!

   MPI_Ssend(&sbuffer[0],length,MPI_DOUBLE,1,tag1,comm);
   MPI_Recv(&sbuffer[0],length,MPI_DOUBLE,1,tag2,comm,&status);
  }

 else if(rank == 1){

   MPI_Recv(&sbuffer[0],length,MPI_DOUBLE,0,tag1,comm,&status);
   MPI_Ssend(&sbuffer[0],length,MPI_DOUBLE,0,tag2,comm);

   }
  }   //end of for loop



  t2 = MPI_Wtime();

  // Extent of an MPI_DOUBLE on this machine

  MPI_Type_size(MPI_DOUBLE,&extent);

  if(rank == 0){
    // Each iteration comprises two messages
    // A MB os 1024*1024 bytes
   totmess = 2.0*extent*length/1024*numiter/1024;
   printf( "Ping-Pong of twice %d bytes, for %d times \n",extent*length,numiter);
   printf( "Total computing time is %f [s]\n", t2 - t1 );
   printf( "Total message size is %f [MB] \n", totmess );
   printf( "Latency (time per message) is %f [s]\n",  0.5*(t2-t1)/numiter);
   printf( "Bandwidth (message size per time) is %f [MB/s]\n",  totmess/(t2-t1));
   
   if((t2-t1)<1.0)
     
     printf( "WARNING: The time is too short to be meaningful, increase the number of iterations and/or the array size so time is at least one second!\n");
  }

  free(sbuffer);

  MPI_Finalize();

}
